src/sinol_make/helpers/printer.py

```diff
@@ -0,0 +1,179 @@
+    curses.curs_set(0)
+    curses.use_default_colors()
+    stdscr.idcok(False)
+    stdscr.idlok(False)
+    stdscr.erase()
+    stdscr.refresh()
+    stdscr.nodelay(True)
+    curses.init_pair(1, curses.COLOR_RED, -1)
+    curses.init_pair(2, curses.COLOR_GREEN, -1)
+    curses.init_pair(3, curses.COLOR_YELLOW, -1)
+
+    curr_row = 0
+    last_output = []
+    last_title = None
+    last_footer = None
+    output = ['']
+    title = None
+    footer = None
+    time = datetime.now()
+    try:
+        while run_event is None or run_event.is_set():
+            inpt = stdscr.getch()
+            height, width = stdscr.getmaxyx()
+
+            if datetime.now() - time > timedelta(seconds=0.1):
+                time = datetime.now()
+                output, title, footer = func(width, height, *args, **kwargs)
+
+            visible_height = height
+            if title is not None:
+                visible_height -= 1
+            if footer is not None:
+                visible_height -= 1
+
+            row_change = False
+            if len(output) > height:
+                if inpt != curses.ERR:
+                    row_change = True
+                    if inpt == curses.KEY_DOWN:
+                        curr_row = min(curr_row + 1, len(output) - visible_height)
+                    elif inpt == curses.KEY_UP:
+                        curr_row = max(curr_row - 1, 0)
+                    elif inpt == curses.KEY_NPAGE:
+                        curr_row = min(curr_row + visible_height, len(output) - visible_height)
+                    elif inpt == curses.KEY_PPAGE:
+                        curr_row = max(curr_row - visible_height, 0)
+                    elif inpt == curses.KEY_END:
+                        curr_row = len(output) - visible_height
+                    elif inpt == curses.KEY_HOME:
+                        curr_row = 0
+                    else:
+                        row_change = False
+
+            if last_output[curr_row:curr_row + visible_height] != output[curr_row:curr_row + visible_height] \
+                    or row_change or last_title != title or last_footer != footer:
+                stdscr.erase()
+
+                if title is not None:
+                    stdscr.addnstr(0, 0, title.ljust(width), width, curses.A_REVERSE)
+                _print_to_scr(stdscr, '\n'.join(output[curr_row:curr_row + visible_height]), title, footer)
+                if footer is not None:
+                    try:
+                        stdscr.addnstr(height - 1, 0, footer.ljust(width), width, curses.A_REVERSE)
+                    except curses.error:  # Curses raises error when trying to write in the lower right corner, but it can be ignored
+                        pass
+
+                stdscr.refresh()
+            last_output = output
+            last_title = title
+            last_footer = footer
+    except KeyboardInterrupt:
+        return
+
+
+def _print_to_scr(scr: curses.window, output, title, footer):
```

Comment:
The code quality of this function is, well, not perfect. As long as we won't need to change this function (or even look at it), I'm OK with merging it, but having some comments or explanations wouldn't hurt, just in case. I leave it up to you -- I'm fine with merging it as it is.

Reply:
I added some comments, what do you think?

Reply:
Yeah, it seems better with those comments, thanks!